lib/remote_node.go

```diff
@@ -0,0 +1,599 @@
+	rn.connectionStatus = RemoteNodeStatus_Validated
+}
+
+// setStatusConnected sets the connection status of the remote node to connected.
+func (rn *RemoteNode) setStatusConnected() {
+	rn.connectionStatus = RemoteNodeStatus_Connected
+}
+
+// setStatusTerminated sets the connection status of the remote node to terminated.
+func (rn *RemoteNode) setStatusTerminated() {
+	rn.connectionStatus = RemoteNodeStatus_Terminated
+}
+
+// setStatusAttempted sets the connection status of the remote node to attempted.
+func (rn *RemoteNode) setStatusAttempted() {
+	rn.connectionStatus = RemoteNodeStatus_Attempted
+}
+
+func (rn *RemoteNode) GetId() RemoteNodeId {
+	return rn.id
+}
+
+func (rn *RemoteNode) GetPeer() *Peer {
+	return rn.peer
+}
+
+func (rn *RemoteNode) getHandshakeMetadata() *HandshakeMetadata {
+	if rn.handshakeMetadata == nil {
+		rn.handshakeMetadata = &HandshakeMetadata{}
+	}
+	return rn.handshakeMetadata
+}
+
+func (rn *RemoteNode) GetNegotiatedProtocolVersion() ProtocolVersionType {
+	meta := rn.getHandshakeMetadata()
+	return meta.negotiatedProtocolVersion
+}
+
+func (rn *RemoteNode) GetValidatorPublicKey() *bls.PublicKey {
+	meta := rn.getHandshakeMetadata()
+	return meta.validatorPublicKey
+}
+
+func (rn *RemoteNode) IsInbound() bool {
+	return rn.peer != nil && !rn.peer.IsOutbound()
+}
+
+func (rn *RemoteNode) IsOutbound() bool {
+	return rn.peer != nil && rn.peer.IsOutbound()
+}
+
+func (rn *RemoteNode) IsPersistent() bool {
+	return rn.peer != nil && rn.peer.IsPersistent()
+}
+
+func (rn *RemoteNode) IsConnected() bool {
+	return rn.connectionStatus == RemoteNodeStatus_Connected
+}
+
+func (rn *RemoteNode) IsValidator() bool {
+	return rn.posValidator
+}
+
+// DialOutboundConnection dials an outbound connection to the provided netAddr.
+func (rn *RemoteNode) DialOutboundConnection(netAddr *wire.NetAddress) error {
+	if rn.connectionStatus != RemoteNodeStatus_NotConnected {
+		return fmt.Errorf("RemoteNode.DialOutboundConnection: RemoteNode is not in the NotConnected state")
+	}
+
+	rn.mtx.Lock()
+	defer rn.mtx.Unlock()
+
+	rn.cmgr.DialOutboundConnection(netAddr, rn.GetId().ToUint64())
+	rn.setStatusAttempted()
+	return nil
+}
+
+// DialPersistentOutboundConnection dials a persistent outbound connection to the provided netAddr.
+func (rn *RemoteNode) DialPersistentOutboundConnection(netAddr *wire.NetAddress) error {
+	if rn.connectionStatus != RemoteNodeStatus_NotConnected {
+		return fmt.Errorf("RemoteNode.DialPersistentOutboundConnection: RemoteNode is not in the NotConnected state")
+	}
+
+	rn.mtx.Lock()
+	defer rn.mtx.Unlock()
+
+	rn.cmgr.DialPersistentOutboundConnection(netAddr, rn.GetId().ToUint64())
+	rn.setStatusAttempted()
+	return nil
+}
+
+// ConnectInboundPeer connects a peer once a successful inbound connection has been established.
+func (rn *RemoteNode) ConnectInboundPeer(conn net.Conn, na *wire.NetAddress) error {
+	if rn.connectionStatus != RemoteNodeStatus_NotConnected {
```

Comment:
I've seen this check a few times. Should we create a function on `connectionStatus` called `IsConnected`?